shared/src/main/scala/mlscript/ucs/Desugarer.scala

```diff
@@ -0,0 +1,605 @@
+import scala.collection.mutable.Buffer
+
+import mlscript._, utils._, shorthands._
+import helpers._
+
+/**
+  * This class contains main desugaring methods.
+  */
+class Desugarer extends TypeDefs { self: Typer =>
+  import Clause.{Conjunction, MatchClass, MatchTuple, BooleanTest}
+
+  type FieldAliasMap = MutMap[Term, MutMap[Str, Var]]
+
+  private var idLength: Int = 0
+
+  def freshName: String = {
+    val res = s"tmp$idLength"
+    idLength += 1
+    res
+  }
+
+  private def desugarPositionals
+    (scrutinee: Term, params: IterableOnce[Term], positionals: Ls[Str])
+    (implicit aliasMap: FieldAliasMap, matchRootLoc: Opt[Loc]): (Buffer[Var -> Term], Ls[Str -> Var]) = {
+    val subPatterns = Buffer.empty[(Var, Term)]
+    val bindings = params.iterator.zip(positionals).flatMap {
+      // `x is A(_)`: ignore this binding
+      case (Var("_"), _) => N
+      // `x is A(value)`: generate bindings directly
+      case (name: Var, fieldName) => S(fieldName -> name)
+      // `x is B(A(x))`: generate a temporary name
+      // use the name in the binding, and destruct sub-patterns
+      case (pattern: Term, fieldName) =>
+        // We should always use the same temporary for the same `fieldName`.
+        // This uniqueness is decided by (scrutinee, fieldName).
+        val alias = aliasMap
+          .getOrElseUpdate(scrutinee, MutMap.empty)
+          .getOrElseUpdate(fieldName, Var(freshName).withLoc(pattern.toLoc))
+        subPatterns += ((alias, pattern))
+        S(fieldName -> alias)
+    }.toList
+    subPatterns -> bindings
+  }
+
+  /**
+    * Desugar sub-patterns from fields to conditions.
+    *
+    * @param subPatterns a list of field name -> pattern term
+    * @param ctx the typing context
+    * @param aliasMap the field alias map
+    * @return desugared conditions representing the sub-patterns
+    */
+  private def destructSubPatterns(subPatterns: Iterable[Var -> Term])
+      (implicit ctx: Ctx, raise: Raise, aliasMap: FieldAliasMap, matchRootLoc: Opt[Loc]): Ls[Clause] = {
+    subPatterns.iterator.flatMap[Clause] {
+      case (scrutinee, subPattern) => destructPattern(scrutinee, subPattern)
+    }.toList
+  }
+
+  private val localizedScrutineeMap = MutMap.empty[Term, Var]
+
+  /**
+    * Create a `Scrutinee`. If the `term` is a simple expression (e.g. `Var` or
+    * `Lit`), we do not create a local alias. Otherwise, we create a local alias
+    * to avoid unnecessary computations.
+    *
+    * @param term the term in the local scrutinee position
+    * @param matchRootLoc the caller is expect to be in a match environment,
+    * this parameter indicates the location of the match root
+    */
+  def makeScrutinee(term: Term, isMultiLineMatch: Bool)(implicit matchRootLoc: Opt[Loc]): Scrutinee =
+    trace(s"Making a scrutinee for $term") {
+      val res = term match {
+        case _: Var => Scrutinee(N, term)
+        case _ =>
+          Scrutinee(
+            S(localizedScrutineeMap.getOrElseUpdate(term, {
```

Comment:
This seems like a pretty bad idea. What of two subterms compare equal for some reason? You'll assign them the same variable! Please find a cleaner and **robust** way to handle this.

Reply:
I'm adding `Loc` to the key of the map. Now it's `MutMap.empty[(Opt[Loc], Term), Var]`.